cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,124 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import javax.annotation.Nullable;
+
+/**
+ * Provides ComputeEngineCredentials either locally if no endpoint is defined, or remotely if endpoint is defined.
+ */
+public final class ComputeEngineCredentials extends GoogleCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME_MILLISECOND = 500;
+  private static final int MAX_WAIT_TIME_MILLISECOND = 10000;
+  private final String endPoint;
+
+  public ComputeEngineCredentials(@Nullable String endPoint) {
+    this.endPoint = endPoint;
+  }
+
+  private AccessToken refreshAccessTokenLocally() throws IOException {
+    try {
+      GoogleCredentials googleCredentials = com.google.auth.oauth2.ComputeEngineCredentials.create();
+      return googleCredentials.refreshAccessToken();
+    } catch (IOException e) {
+      throw new IOException("Unable to get credentials from the environment. "
+                              + "Please explicitly set the account key.", e);
+    }
+  }
+
+  private AccessToken refreshAccessTokenRemotely(String endPoint) throws IOException {
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME_MILLISECOND)
+      .setMaxIntervalMillis(MAX_WAIT_TIME_MILLISECOND).build();
+
+    Exception exception = null;
+    int counter = 1;
+    while (counter <= NUMBER_OF_RETRIES) {
+      try {
+        return fetchAccessToken(endPoint);
+      } catch (Exception ex) {
+        // exception does not get logged since it might get too spammy.
+        exception = ex;
+      }
+
+      try {
+        Thread.sleep(backOff.nextBackOffMillis());
+      } catch (InterruptedException ex) {
+        exception = ex;
+        break;
+      }
+      counter++;
+    }
+
+    LOG.error("Failed to fetch GoogleCredentials after {} retries", counter, exception);
+    return null;
+  }
+
+  private AccessToken fetchAccessToken(String endPoint) throws IOException {
+    URL url = new URL(endPoint);
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    connection.connect();
+    try (Reader reader = new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8)) {
+      if (connection.getResponseCode() != HttpResponseStatus.OK.code()) {
+        throw new IOException(CharStreams.toString(reader));
+      }
+      GenericData token = GSON.fromJson(reader, GenericData.class);
+
+      String key = token.get("access_token").toString();
```

Comment:
Both the key and the expiration can potentially be null if for any reason the json blob changed. I think it's better to do a field existence check before using it.